[WIP] [#847] : Implement push notification server with support for gorush

adriacidre opened the pull request:
**This is an in progress pull request so the team behind [Push Notification Epic](https://github.com/status-im/status-go/issues/846) can synchronize its work**

Push notification server to be a bridge between whisper and [gorush](https://github.com/appleboy/gorush).

Important changes:
- [ ] ...

Closes #847

adriacidre committed:
[#847] : Implement push notification server with support for gorush

adriacidre commented:
@PombeirP tomorrow i'll be exploring how to merge logic on [notifier.go](https://github.com/status-im/status-go/pull/863/files#diff-0bbaa14dc7b2b9d080e12d3716bc6f2f) with the current [fcm logic](https://github.com/status-im/status-go/tree/develop/geth/notifications/push) and see if we still need that. 

I think we can split the work from this point, so you can focus on making the server able to read whisper messages.

pedropombeiro commented:
@adriacidre we'll need to replace the current FCM API (both on status-go and status-react) with a simpler API that will be called from https://github.com/status-im/status-go/blob/develop/geth/api/backend.go#L420 and simply send the Whisper message to the PN server (assuming the negotiation has already taken place):

![image](https://user-images.githubusercontent.com/138074/39202807-54a3577c-47f3-11e8-8ac9-d084e4fa1bfe.png)


<blockquote><img src="https://avatars3.githubusercontent.com/u/11767950?s=400&v=4" width="48" align="right"><div><img src="https://assets-cdn.github.com/favicon.ico" height="14"> GitHub</div><div><strong><a href="https://github.com/status-im/status-go">status-im/status-go</a></strong></div><div>status-go - The Status module that consumes go-ethereum</div></blockquote>

pedropombeiro commented on cmd/notifier/main.go:
```diff
@@ -0,0 +1,51 @@
+	prodMode            = flag.Bool("production", false, "Whether production settings should be loaded")
+	dataDir             = flag.String("datadir", params.DataDir, "Data directory for the databases and keystore")
+	networkID           = flag.Int("networkid", params.RopstenNetworkID, "Network identifier (integer, 1=Homestead, 3=Ropsten, 4=Rinkeby, 777=StatusChain)")
+	pushNotificationURI = flag.String("uri", "localhost:3000", "Data directory for the databases and keystore")
```
Copy/pasted help string

pedropombeiro commented on cmd/notifier/notifier.go:
```diff
@@ -0,0 +1,47 @@
+// Send : Sends a push notification to given devices
+func (n *Notifier) Send(tokens []string, message string) error {
+	_, err := n.c.Send(context.Background(), &proto.NotificationRequest{
+		Platform: 2,
```
Hopefully we can continue to use the FCM platform (2) to send notifications to both Android and iOS, otherwise we'll need to store this value at registration time.

adriacidre committed:
Introduced status messaging library and started broadcast availability

adriacidre commented:
@PombeirP i've introduced the necessary libraries to interact with all new protocol message types `~#c1, ~#c2, ~#c3, ~#c4, ~#c5, ~#c6`. This command should be broadcasting availability now.

From now on i'll be implementing the ability to subscribe to different message types as this is only supporting standard message type `~#c4`.

The code still has lots of TODOs and it's not finished, but you can have a look around so we can agree on the datastructures :)

adriacidre commented on messenger/conn.go:
```diff
@@ -0,0 +1,197 @@
+package messenger
```
probably not the best name in the world, i'm open to suggestions here

pedropombeiro commented on cmd/notifier/messenger.go:
```diff
@@ -0,0 +1,135 @@
+// NotificationRequestMsg : a registered contact requests sending a push
+// notification to one of its contacts
+type NotificationRequestMsg struct {
+	// TODO (adriacidre) : Check @PombeirP what fields are needed here
```
we'll need:
- the message ID, which is the hash of the whisper message (sent from A->B) payload
- the destination contact's AK2 public key

pedropombeiro commented on cmd/notifier/messenger.go:
```diff
@@ -0,0 +1,135 @@
+// securely
+type RegistrationRequestMsg struct {
+	DeviceRegistrationToken string `json:"token"`
+	SecureChannel           []byte `json:"channel"`
```
The channel is composed of the symkey and a topic. I think we're missing the Whisper topic here.

pedropombeiro commented on cmd/notifier/messenger.go:
```diff
@@ -0,0 +1,135 @@
+)
+
+type BroadcastMsg struct {
+	PubKey string `json:"pubkey"`
```
@adriacidre Where is this being used? I'd expect it to be used in `BroadcastAvailability` to pass the pubkey to the clients (which they will use to encrypt their response to us).

adriacidre replied:
I'll remove it, it was like this. But that logic was moved to the messenger library. 

Now this will happen through the [NewContactKeyRequest](https://github.com/status-im/status-go/pull/863/files#diff-e0aa9266aa1e17a3ee515e4d8af3f93eR81) as described on the  New status protocol doc:
```
[
  "~#c1",
  [
    "0xc2a38488c043b7c0d138e922887bfcc2966c7084....",
    "0x2d215ad8",
    [
      "~#c2",
      [
        "my name",
        "data:image/png;base64,base64-representation-of-profile-image...",
        "e5ac45596d8be7aabea51c0dbbac34.....",
        "dFLRZwPx3yQ:APA91bH_7yjq47N5HY0ciKs58HW3VbJ9...”
      ]
    ]
  ]
]
```

pedropombeiro commented on cmd/notifier/messenger.go:
```diff
@@ -0,0 +1,135 @@
+func (m *Messenger) ManageRegistrations() {
+	log.Println("Subscribed to discovery topic :", m.discoveryTopic)
+	filterID, _, _ := m.client.JoinPublicChannel(m.discoveryTopic)
+	for range time.Tick(m.pollInterval) {
```
@adriacidre Do we need to periodically call `Subscribe`??

adriacidre replied:
Yes, I think that's how it works, and its a pita in terms of performance (as we will have n topic to be subscribed).

On the other hand I think you can subscribe to multiple topics at the same time, so we can allow subscription object to manage the topics is subscribed to.

pedropombeiro replied:
From the [source code](https://github.com/status-im/status-go/blob/feature/issue-847/gorush-based-push-notification-server/messenger/subscription.go#L33) it looks like it will wait for messages indefinitely, so I don't see why we need to call `Subscribe` over and over again with the same `filterID`...

adriacidre replied:
Ah, damn it, you're right. Subscribe is already implementing this loop, it doesn't need to exist here

adriacidre commented:
## PN Negotiation process

### 0 .- Pre-conditions

- [ ] The PN is subscribed to the default discovery topic 
- [ ] All clients are subscribed default discovery topic 

### 1 .- PN to broadcast a message on a specific discovery topic

PN has to broadcast its public key so anybody in the network can start an interaction with him. There is no message type matching this spec (BroadCastAvailability) on the current protocol. Following the current protocol a proposal could look like:
```
[
  "~#cN",
  [
    "843bacc6c484e622da922aab3eaa08ac568d9986", <---- address (PN public key)
  ]
]

```

#### Implementation notes

- [ ] New message type has to be defined
- [ ] New message type will be emited periodically from [here](https://github.com/status-im/status-go/pull/863/files#diff-e0aa9266aa1e17a3ee515e4d8af3f93eR81) instead NewContactKeyRequest



### 2 .- A contact (C1) receives a `BroadCastAvailability` message type over discovery topic

Any client on the network has to be listening for messages on the `discovery topic`. When a message type `BroadCastAvailability` enters, it should:

- [ ] Generate a random secure topic (secure topic) based on random symmetric key + topic.
- [ ] Subscribe to secure topic for `ContactRequestConfirmed` messages.
- [ ] Send a `NewContactKey` message providing its Public key and the proposed `secure topic`, through the `discovery topic`, this message like:
```
[
  "~#c1",
  [
    "0xc2a38488c043b7c0d138e922887bfcc2966c70849774b6d815596737cc654ee1", // <---- symetric key (SK1)
    "0x2d215ad8", // <---- secure channel
    [
      "~#c2",
      [
        "",  // <---- contact name
        ".",  // <---- image
        "e5ac45596d8be7aabea51c0dbbac3494e6a1636d",    // <----- (SK1)
        "dFLRZwPx3yQ:APA91bH_7yjq47N5HY0ciKs58HW3VbJ9..." // <---- device registration token
      ]
    ]
]
```

This last field is currently being used to send `fcm-token`, we have to re-use it to send  `device registration token` instead.


#### Implementation notes

The `messenger` package exposes a [helper method](https://github.com/status-im/status-go/pull/863/files#diff-7c61f41a81252ceffdb0c8961606a875R132) to send `NewContactKey` rpc requests.


### 3 .- PN receives a `NewContactKey` msg over `discovery topic`

The PN has to be listening on the discovery topic for `NewContactKey` messages, once it receives one, it will:

- [ ] joins the proposed secure topic and subscribe to all its messages.
- [ ] respond with a `ContactRequestConfirmed`
So, this response will look like:
```
[
  "~#c3",
  [
    "PN server",												<---- Username
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAA...",		<---- image
    "843bacc6c484e622da922aab3eaa08ac568d9986",					<---- address (PN public key)
    ""															<---- device registration token (not needed)
  ]
]
```

#### Implementation notes

- `messenger` package provides a [helper method](https://github.com/status-im/status-go/pull/863/files#diff-7c61f41a81252ceffdb0c8961606a875R156) to send ConfirmedContact Requests


### 4 .- C1 receives the `ContactRequestConfirmed` through `secure topic`

- [ ] Store (inmemory?) the PN public key / topic for future communications


### 5 .- C1 sends a push notification request

No message like this exists on the current protocol, we should define a new one (NotificationRequest) to avoid collisions with existing ones 
```
[
  "~#cN",
  [
    "843bacc6c484e622da922aab3eaa08ac568d9986", <---- address (Contact B public key)
  ]
]
```
- [ ] Define new message
- [ ] Contacts to send this message to in memory PNs
- [ ] PN to listen and process NotificationRequest messages


### Random concerns

#### What does it happen if you have multiple PN on the network? 
You can potentially be talking to a PN, but your contact can be registered on a different one :/

#### Security
Anybody adding a PN to the network could get all device registration tokens :/

cc: @PombeirP

pedropombeiro commented:
@adriacidre , in 2. we don't want to send the real public key for C2. We want to maintain as much anonymity as possible, so that the PN doesn't even keep the association between contact pubkey <-> device registration token. The idea is that the PN server will create AK2, return it to C1, and store that associated with C1's device registration token.

pedropombeiro commented:
> This last field is currently being used to send fcm-token, we have to re-use it to send device registration token instead.

In the future when we support other PN platforms such as Pushy, we'll need to know the type of the token we're describing. Maybe it'll just be added as a prefix to the value, e.g. (`pushy://<device-token>`)

pedropombeiro commented:
> Store (inmemory?) the PN public key / topic for future communications

C1 should store pubkey of AK2 (which it received from PN server). It should be stored durably as this is something C1 will pass to C2 so that C2 knows how to ask the PN server to send PNs to C1.

> You can potentially be talking to a PN, but your contact can be registered on a different one :/

I think we'll need to send not only AK2's pubkey in `NewContactKey`/`ContactRequestConfirmed`, but also the respective PN server's AK1 pubkey, so that the sender can ask the correct PN server to notify its client.

> Anybody adding a PN to the network could get all device registration tokens :/

The user will be able to choose which PN server he trusts when there is more than one server available. This means that the PN server will only have the device registration tokens for the devices that decided to register with it. He won't however be able to tell which Status account matches a device registration token. As an improvement, we could change Gorush so that it makes use of an asymmetric key to decrypt the device registration token.

cc @yenda

adriacidre committed:
Subscription supporting multiple message types

pedropombeiro commented on messenger/subscription.go:
```diff
@@ -0,0 +1,117 @@
+	}
+}
+
+// Subscribe : Listens to a specific topic and executes the given function for
```
@adriacidre There's no "given" function here (nothing is passed as an argument to `Subscribe`)...

Commit:
Move testable code from cmd/notifier to its own package (temporary until we use the Go SDK package).

- Run dep-ensure
- Fix race condition pointed out by `go test -race` in `messenger/subscription.go`
- Fix topic name being passed instead of topic ID
- Make `Messenger` methods return an error parsed from the JSON response

Commit:
Fix a couple of bugs preventing the PN server from successfully sending messages

Commit:
Add initial test for PN subscription process

adriacidre committed:
Introduced status-go-sdk

adriacidre committed:
new TODOS and support for private channels

adriacidre committed:
NotificationProvider interface introduced to make testing easy

adriacidre committed:
Closed integration with last sdk api

Commit:
Adapt e2e test

adriacidre committed:
Remove multiple account dependency

adriacidre committed:
Latest sdk

adriacidre committed:
Updat sdk to last version

adriacidre committed:
Client commented logic as e2e test

adriacidre committed:
Removed gorush package dependency

status-im-jenkins commented:
Can one of the admins verify this patch?

dshulyak commented:
@PombeirP i will close it. please re-open if it needs to be open for some reason :)

ghost commented:
<!--prchecklist--> 
### Pull Request Checklist
- [ ] Have you updated the documentation, if impacted (e.g. [docs.status.im](https://docs.status.im/docs/build_status_go.html))?
